File: 0189-rotate-array/0189-rotate-array.cpp
class Solution {
private:
    void reverse(vector<int>& nums, int start, int end) {
        while(start < end) {
            int temp = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            start++;
            end--;
        }
    }
public:
    void rotate(vector<int>& nums, int k) {
        int n = nums.size();
        k %= n; // handling out of bound exception when k > n
        reverse(nums, 0, n - 1); // reversing the whole array
        reverse(nums, 0, k - 1); // reversing that array upto kth index
        reverse(nums, k, n - 1); // reversing that array from kth + 1 index upto the end
    }
};

File: 0189-rotate-array/0189-rotate-array.java
class Solution {
    public static void reverse(int i, int j, int[] nums) {
        while(i < j) {
            int temp = nums[i];
            nums[i++] = nums[j];
            nums[j--] = temp;
        }
    }

    public void rotate(int[] nums, int k) {
        int n = nums.length;
        k %= n;
        reverse(0, n - 1, nums);
        reverse(0, k - 1, nums);
        reverse(k, n - 1, nums);
    }
}